lib/preview-web/src/PreviewWeb.tsx

```diff
@@ -0,0 +1,559 @@
+    >;
+    element: Element;
+  }) {
+    const { id, applyLoaders, unboundStoryFn, runPlayFunction } = story;
+
+    // IE11 doesn't support AbortController, so we either need to polyfill or just not support it
+    const controller = AbortController
+      ? new AbortController()
+      : {
+          signal: { aborted: false },
+          abort() {
+            this.signal.aborted = true;
+          },
+        };
+    let initialRenderPhase: InitialRenderPhase = 'init';
+    let renderContext: RenderContext<TFramework>;
+    const initialRender = async () => {
+      const storyContext = this.storyStore.getStoryContext(story);
+
+      const { parameters, initialArgs, argTypes, args } = storyContext;
+      if (FEATURES?.storyStoreV7) {
+        this.channel.emit(Events.STORY_PREPARED, {
+          id,
+          parameters,
+          initialArgs,
+          argTypes,
+          args,
+        });
+      }
+
+      const viewMode = element === this.view.storyRoot() ? 'story' : 'docs';
+      const loadedContext = await applyLoaders({
+        ...storyContext,
+        viewMode,
+      } as StoryContextForLoaders<TFramework>);
+      if (controller.signal.aborted) {
+        return;
+      }
+      initialRenderPhase = 'loaded';
+
+      // By this stage, it is possible that new args/globals have been received for this story
+      // and we need to ensure we render it with the new values
+      const updatedStoryContext = {
+        ...loadedContext,
+        ...this.storyStore.getStoryContext(story),
+      };
+      renderContext = {
+        ...renderContextWithoutStoryContext,
+        // Whenever the selection changes we want to force the component to be remounted.
+        forceRemount: true,
+        storyContext: updatedStoryContext,
+        storyFn: () => unboundStoryFn(updatedStoryContext),
+        unboundStoryFn,
+      };
+      await this.renderToDOM(renderContext, element);
+
+      if (controller.signal.aborted) {
+        return;
+      }
+      initialRenderPhase = 'rendered';
+
+      // NOTE: if the story is torn down during the play function, there could be negative
+      // side-effects (as the play function tries to modify something that is no longer visible).
+      // In the future we will likely pass the AbortController signal into play(), and also
+      // attempt to scope the play function by passing the element.
+      //
+      // NOTE: it is possible that args/globals have changed in between us starting to render
+      // the story and executing the play function (it is also possible that they change mid-way
+      // through executing the play function). We explicitly allow such changes to re-render the
+      // story by setting `initialRenderDone=true` immediate after `renderToDOM` completes.
+      await runPlayFunction();
+      if (controller.signal.aborted) {
+        return;
+      }
+      initialRenderPhase = 'done';
+
+      this.channel.emit(Events.STORY_RENDERED, id);
```

Comment:
Reminder from what was discussed with Norbert already: maybe STORY_RENDERED should be done before the playFunction and new event should be emitted instead?

Reply:
I don't think we can do this as it is a breaking change (although I'd like to). We could consider doing it behind the `breakingChangeV7` flag. @ghengeveld and I discussed a new event (`STORY_PHASE_REACHED` or something).

Reply:
Having `STORY_RENDERED` at the end is nice because tools like Chromatic get `play` function support transparently. To me keeping `STORY_RENDERED` at the end and introducing a more generic `STORY_PHASE` event sounds like a practical and general direction.


That said, I don't think it's really a breaking change if we wanted to move `STORY_RENDERED` earlier in the render process. Technically speaking, `play` is still an experimental feature and we could conceivably move it after `STORY_RENDERED` and introduce a `STORY_PLAYED` event that really marks the end of the render cycle, or maybe a more future proof `STORY_DONE` event. Not proposing we do this, but if we did I think it'd be OK from a semver perspective.